libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/Edge.kt

```diff
@@ -0,0 +1,407 @@
+package com.mapbox.navigation.base.trip.model
+
+import com.mapbox.geojson.LineString
+
+/**
+ * Basic Edge class
+ *
+ * @param id the Edge identifier
+ * @param level the level of the Edge (0 being the mpp, 1 branches of the mpp,
+ * 2 branches of level 1 branches, etc)
+ * @param probability the probability of the Edge in percentage
+ * @param heading the bearing in degrees clockwise at the start of the edge
+ * @param length the Edge's length in meters
+ * @param out the outgoing Edges
+ * @param frc the edge's functional road class
+ * @param wayId the edge's way ID
+ * @param positiveDirection Specifies if the geometry is in the forward or
+ * backwards direction
+ * @param speed the Edge's average speed
+ * @param ramp is the edge a ramp?
+ * @param motorway is the edge a motorway?
+ * @param bridge is the edge a bridge?
+ * @param tunnel is the edge a tunnel?
+ * @param toll is the edge a toll road?
+ * @param names the edge's names
+ * @param curvature the edge's curvature
+ * @param geometry optional geometry if requested.
+ * @param speedLimit the Edge's max speed
+ * @param laneCount the edge's lane counts
+ * @param meanElevation the edge's mean elevation
+ * @param countryCode the edge's country code
+ * @param stateCode the edge's state code
+ */
+class Edge private constructor(
+    val id: Long,
+    val level: Byte,
+    val probability: Double,
+    val heading: Double,
+    val length: Double,
+    val out: List<Edge>,
+    val frc: String,
+    val wayId: String,
+    val positiveDirection: Boolean,
+    val speed: Double,
+    val ramp: Boolean,
+    val motorway: Boolean,
+    val bridge: Boolean,
+    val tunnel: Boolean,
+    val toll: Boolean,
+    val names: List<NameInfo>,
+    val curvature: Byte,
+    val geometry: LineString?,
+    val speedLimit: Double?,
+    val laneCount: Byte?,
+    val meanElevation: Double?,
+    val countryCode: String?,
+    val stateCode: String?
+) {
+
+    /**
+     *
+     */
+    fun <R> accept(visitor: EdgeVisitor<R>): R {
+        return visitor.visit(this)
+    }
+
+    /**
+     *
+     */
+    fun <R> collect(fn: (edge: Edge) -> R?): List<R> {
+        class Visitor : EdgeVisitor<List<R>> {
+
+            override fun visit(edge: Edge): List<R> {
+                val result: R? = fn(edge)
+                return if (result != null) {
+                    listOf(result)
+                } else {
+                    emptyList()
+                } + edge.out.flatMap { outEdge -> outEdge.accept(this) }
+            }
+        }
+        return this.accept(Visitor())
+    }
```

Comment:
What's the purpose of these functions? Is it a utility to invoke a function (transformation) on all nested edges?

Why not expose only one function (more Java friendly as well):
```
class Edge {
  fun <R> transformAll(transformation: EdgeTransformation<R>): List<R> { ... }
}
```
I feel like it would deliver the same result without the need to first acquire the visitor and then executing it.